xds/server.go

```diff
@@ -233,6 +251,116 @@ func (s *GRPCServer) newListenerWrapper(lis net.Listener) (*listenerWrapper, err
 	return lw, nil
 }
 
+// watchUpdate wraps the information received from a registered LDS watcher.
+type watchUpdate struct {
+	lw         *listenerWrapper         // listener associated with this watch
+	name       string                   // resource name being watched
+	lds        xdsclient.ListenerUpdate // received update
+	err        error                    // received error
+	goodUpdate *grpcsync.Event          // event to fire upon a good update
+}
+
+func (s *GRPCServer) handleListenerUpdate(update watchUpdate) {
+	if update.lw.closed.HasFired() {
+		s.logger.Warningf("Resource %q received update: %v with error: %v, after for listener was closed", update.name, update.lds, update.err)
+		return
+	}
+
+	if update.err != nil {
+		// We simply log an error here and hope we get a successful update
+		// in the future. The error could be because of a timeout or an
+		// actual error, like the requested resource not found. In any case,
+		// it is fine for the server to hang indefinitely until Stop() is
+		// called.
+		s.logger.Warningf("Received error for resource %q: %+v", update.name, update.err)
+		return
+	}
+	s.logger.Infof("Received update for resource %q: %+v", update.name, update.lds.String())
+
+	if err := s.handleSecurityConfig(update.lds.SecurityCfg, update.lw); err != nil {
+		s.logger.Warningf("Invalid security config update: %v", err)
+		return
+	}
+
+	// If we got all the way here, it means the received update was a good one.
+	update.goodUpdate.Fire()
+}
+
+func (s *GRPCServer) handleSecurityConfig(config *xdsclient.SecurityConfig, lw *listenerWrapper) error {
+	// If xdsCredentials are not in use, i.e, the user did not want to get
+	// security configuration from the control plane, we should not be acting on
+	// the received security config here. Doing so poses a security threat.
+	if !s.xdsCredsInUse {
+		return nil
```

Comment:
Should we return an error?  Or at least log something?  (Except I guess this is expected if `config==nil`, right?)

Reply:
Aha, sorry I saw "poses a security threat" and thought that meant we had a near-miss.